python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,753 @@
+
+
+class DLPackObject:
+    def __init__(self, value) -> None:
+        try:
+            self._capsule = _dlpack.get_dlpack_capsule(value)
+            self._tensor = _dlpack.get_managed_tensor(self._capsule).dl_tensor
+        except Exception as e:
+            raise InvalidArgumentError(
+                f"Object does not support DLPack protocol: {e}"
+            ) from None
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, DLPackObject):
+            return False
+        if self.byte_size != other.byte_size:
+            return False
+        if self.memory_type != other.memory_type:
+            return False
+        if self.memory_type_id != other.memory_type_id:
+            return False
+        if self.shape != other.shape:
+            return False
+        if self.data_ptr != other.data_ptr:
+            return False
+        if self.contiguous != other.contiguous:
+            return False
+        if self.triton_data_type != other.triton_data_type:
+            return False
+        return True
+
+    @property
+    def byte_size(self) -> int:
+        return _dlpack.get_byte_size(
+            self._tensor.dtype, self._tensor.ndim, self._tensor.shape
+        )
+
+    @property
+    def memory_type(self) -> MemoryType:
+        return DLPACK_DEVICE_TYPE_TO_TRITON_MEMORY_TYPE[self._tensor.device.device_type]
+
+    @property
+    def memory_type_id(self) -> int:
+        return self._tensor.device.device_id
+
+    @property
+    def shape(self) -> list[int]:
+        return [self._tensor.shape[i] for i in range(self._tensor.ndim)]
+
+    @property
+    def triton_data_type(self) -> DataType:
+        return DLPACK_TO_TRITON_DTYPE[self.data_type]
+
+    @property
+    def data_type(self) -> tuple[_dlpack.DLDataTypeCode, int]:
+        return (self._tensor.dtype.type_code, self._tensor.dtype.bits)
+
+    @property
+    def data_ptr(self) -> ctypes.c_void_p:
+        return self._tensor.data + self._tensor.byte_offset
+
+    @property
+    def contiguous(self) -> bool:
+        return _dlpack.is_contiguous_data(
+            self._tensor.ndim, self._tensor.shape, self._tensor.strides
+        )
+
+
+@dataclass
+class Tensor:
+    data_type: DataType
+    shape: Sequence[int]
+    memory_buffer: MemoryBuffer
+
+    @property
+    def data_ptr(self) -> int:
+        return self.memory_buffer.data_ptr
+
+    @property
+    def memory_type(self) -> MemoryType:
+        return self.memory_buffer.memory_type
+
+    @property
+    def memory_type_id(self) -> int:
+        return self.memory_buffer.memory_type_id
+
+    @property
+    def size(self) -> int:
+        return self.memory_buffer.size
+
+    def __dlpack__(self, *, stream=None):
+        #        if not (stream is None or (isinstance(stream, int) and stream == 0)):
+        #           raise UnsupportedError(
+        #              f"DLPack stream synchronization on {stream} not currently supported"
+        #         )
```

Comment:
Let's remove the comment.

Reply:
added todo - and discussing with @oandreeva-nv